src/endpoint_connection.rs

```diff
@@ -12,72 +12,77 @@
 //  See the License for the specific language governing permissions and
 //  limitations under the License.
 
-use std::{
-    os::unix::net::{UnixStream, UnixListener},
-    os::unix::io::{RawFd, AsRawFd},
-    path::Path,
-    fs,
-};
+use anyhow::{Result, Context};
 use crate::{
     criu,
-    util::{pb_write, recv_fd, pb_read_next},
     unix_pipe::{UnixPipe, UnixPipeImpl},
+    util::{pb_write, recv_fd, pb_read_next, pb_read_next_breaking},
+};
+use std::{
+    fs,
+    os::unix::fs::PermissionsExt,
+    os::unix::io::{RawFd, AsRawFd},
+    os::unix::net::{UnixStream, UnixListener},
+    path::Path,
 };
-use anyhow::{Result, Context};
 
-/// The role of the `CriuListener` and `CriuConnection` is to handle communication with CRIU and
-/// Cedana over the image sockets.
+/// The role of the `EndpointListener` and `EndpointConnection` is to handle communication with 
+/// endpoints cedana, cedana-gpu-controller, and criu over the image sockets.
 
-pub struct CriuListener {
+pub struct EndpointListener {
     listener: UnixListener,
 }
 
-impl CriuListener {
-    fn bind(socket_path: &Path) -> Result<Self> {
+impl EndpointListener {
+    pub fn bind(images_dir: &Path, socket_name: &str) -> Result<Self> {
         // 1) We unlink the socket path to avoid EADDRINUSE on bind() if it already exists.
         // 2) We ignore the unlink error because we are most likely getting a -ENOENT.
         //    It is safe to do so as correctness is not impacted by unlink() failing.
-        let _ = fs::remove_file(socket_path);
-        let listener = UnixListener::bind(socket_path)
+        let socket_path = images_dir.join(socket_name);
+        let _ = fs::remove_file(&socket_path);
+        let listener = UnixListener::bind(&socket_path)
             .with_context(|| format!("Failed to bind socket to {}", socket_path.display()))?;
 
-        Ok(Self { listener })
-    }
+        // Adjust permissions to allow cedana-gpu-controller to connect
+        let mut permissions = fs::metadata(&socket_path)?.permissions();
+        permissions.set_mode(0o777);
```

Comment:
Do we need the execute permission? Consider 0o666?

Reply:
good catch!